[6.4] Docs: Add 6.4 release notes and other changes (#9851)
apple/swift-package-manager: Sources/PackageManagerDocs/Documentation.docc/SwiftBuildDefault.md

# [6.4] Swift Build System now the default in SwiftPM

Understand and use the next-generation build system for Package Manager.

@Metadata {

>     @Available("Swift", introduced: "6.3")

Reviewer: 6.4?

Author: I'm not sure. SwiftBuild was introduced in 6.3.  I'm not sure what this metadata tag does. 

@heckj: Any thoughts on what the value should be here?

}

## Overview

> Swift Package Manager previewed [Swift Build](https://github.com/swiftlang/swift-build) as a replacement for the current native build system in Swift 6.3, and it became the default build system in Swift 6.4. This document outlines the key differences and known issues.

Reviewer: how about:

```
Swift Package Manager introduced [Swift Build](https://github.com/swiftlang/swift-build) as a preview replacement for its native build system in Swift 6.3 and adopted it as the default in Swift 6.4. This document outlines the key differences and known issues.
```

> We encourage you to [try](#How-to-use-the-Swift-Build-build-system) the build system in your project and [report any issues](#Reporting-issues) you encounter.

Reviewer: I think "use the new build system" would be better than "try the new build system", since "try" feels like it implies an opt-in, but it's now opt-_out_.

Author: 100% agree.

> For more information about the migration to Swift Build, see [this forum post](https://forums.swift.org/t/evolving-swiftpm-builds-with-swift-build/77596).

Reviewer: How about:

```
For more information about the migration to Swift Build, see [Evolving SwiftPM Builds with Swift Build](https://forums.swift.org/t/evolving-swiftpm-builds-with-swift-build/77596) on the Swift Forums.
```

(using "this" or "here" in hyperlinks is considered bad practice for accessibility)

Author: Sounds good.

> For more information about the migration to Swift Build, see [Evolving SwiftPM buildswith Swift Build](https://forums.swift.org/t/evolving-swiftpm-builds-with-swift-build/77596) on the [Swift Forums](https://forums.swift.org).

Reviewer 2: The framing on this doc is a mixture of upsell, migration information content, and gaps - and it feels more like a blog post (a point in time status update kind of thing) that documentation that's expected to be consistent and live for an extended period.

We should find and pull up any content from the forums that are linked above rather than directing someone into the forums for that information. This should (ideally) be the refinement of what the migration process it, and it's what I expected I might find based on the title.

There's a little bit of awkward framing as "native" vs. "next generation" that's also very point-in-time focused. If there's something you need to do differently to build a swift package with the new swift-build as a default - then we should be updating that docs kind of across the board where-ever that's relevant, and I'd suggest framing this article as a migration guide - if we need one.

For the known issues and gaps, and reporting issues, I'd keep all of that entirely with the release notes article. The earlier analog to this was "here's a preview, please try it", but with that complete any docs content should be focused more on "here's how it works" vs. "here's what's new". The "here's what's new" is exactly what release notes are good at, as they're expected to be reflecting a specific point in time.

## Migration plan

The Swift Build integration follows these phases:

1. **Preview Phase** - Testing and feedback collection (Swift 6.3)
2. **Feature Parity** - Address remaining gaps and platform issues
3. **Default Migration** - Transition to Swift Build as default (Swift 6.4 - Current)

> 4. **Legacy Build System Deprecation** - Phase out `native` and `xcode` build systems

Reviewer: This should say "removal", not "deprecation". `native` and `xcode` are already deprecated as of Swift 6.4.

Author: Done.

## How to use the Swift Build build system

As of Swift 6.4, Swift Build is the default build system and no additional flags are required:

```bash
swift build
swift test
swift run
```
### Using Swift Build in Swift 6.3

In Swift 6.3 (preview phase), you need to explicitly opt-in using the `--build-system swiftbuild` flag:

```bash
swift build --build-system swiftbuild
swift test --build-system swiftbuild
swift run --build-system swiftbuild
```

## Key improvements and differences between Native build system

> - Build artifacts are output to a different location.  To determine the build output location, use the `swift build --show-bin-path <other build arguments>`

Reviewer: worth clarifying that the --show-bin-path is best practice for SB _and_ native build systems

Author: Great feedback. I updated the wording  to indicate `--show-bin-path` is the recommended way for all build systems.

- **`--static-swift-stdlib`**: Stricter validation (errors instead of silently ignoring)
  - The native build system silently ignored this option on some platforms; Swift Build now produces an error.

>   - The Swift Build build system will generate an error on platform tht do not support static libraries are supported.

Reviewer: ```suggestion
  - The Swift Build build system will generate an error on platforms where static libraries are not supported.
```

Author: Completed.  thanks for the feedback.

>     See [Upcoming changes to Windows Swift SDKs](https://forums.swift.org/t/upcoming-changes-to-windows-swift-sdks/81313).

Reviewer: It's not clear how this article is related to the error checking.

Author: The article was meant to provide information on whether Swift static stdlib is supported on Windows. 

I will remove the reference.

    ```
- Swift Build tasks planned by build tool plugins no longer have unrestricted access to environment variables in order
  to avoid excessive invalidation in incremental builds.

### Detailed differences

#### Resource support parity with xcodebuild

> - When targeting Apple platforms, SwiftPM, by default, consistently applies the same set of resource rules as `xcodebuild` when a package contains resource types like asset catalogs, storyboards, Metal sources, etc.

Reviewer: ```suggestion
- When targeting Apple platforms, the same set of resource rules that `xcodebuild` follows are consistently applied when a package contains resource types like asset catalogs, storyboards, Metal sources, etc.
```

Author: Should be updated. thanks.

#### Swift Driver integration

> - The `--use-integrated-swift-driver` command-line option is considered deprecated. The default now always uses the library-based Swift driver.

Reviewer: ```suggestion
- The `--use-integrated-swift-driver` command-line option is deprecated and has no effect. Swift Build always uses the library-based Swift driver.
```

- On supported platforms, SwiftPM will make use of explicitly-built Clang and Swift modules.

#### Enhanced diagnostics
- Logging and diagnostic differences between native and Swift Build build systems

#### Other

- Supports building universal binaries when targeting Apple platforms. Example invocation: `swift build --arch arm64 --arch x86_64`

## Known issues

- Coverage reporting issues on some platforms (Linux, Windows)
  - **Tracking**: [swiftlang/swift-package-manager#9600](https://github.com/swiftlang/swift-package-manager/issues/9600)

> - SDK Generator fails on Linux

Reviewer: Did you mean to say "Swift Build fails to build projects with Linux SDKs generated _by_ the SDK generator"?

Author: yes. I had left that there as a placeholder.
